src/function/table/table_scan.cpp

```diff
@@ -361,56 +362,99 @@ unique_ptr<GlobalTableFunctionState> DuckIndexScanInitGlobal(ClientContext &cont
 	return std::move(g_state);
 }
 
-void ExtractInFilter(unique_ptr<TableFilter> &filter, BoundColumnRefExpression &bound_ref,
-                     unique_ptr<vector<unique_ptr<Expression>>> &filter_expressions) {
-	// Special-handling of IN filters.
-	// They are part of a CONJUNCTION_AND.
-	if (filter->filter_type != TableFilterType::CONJUNCTION_AND) {
-		return;
-	}
+void ExtractExpressionsFromValues(vector<Value> &values, BoundColumnRefExpression &bound_ref,
+                                  unique_ptr<vector<unique_ptr<Expression>>> &expressions) {
+	// Ensure that we never push duplicates.
+	value_set_t unique_values;
 
-	auto &and_filter = filter->Cast<ConjunctionAndFilter>();
-	auto &children = and_filter.child_filters;
-	if (children.empty()) {
-		return;
+	// They are all on the same column, so we can split them.
+	for (const auto &value : values) {
+		if (unique_values.find(value) != unique_values.end()) {
+			continue;
+		}
+		auto bound_constant = make_uniq<BoundConstantExpression>(value);
+		auto filter_expr = make_uniq<BoundComparisonExpression>(ExpressionType::COMPARE_EQUAL, bound_ref.Copy(),
+		                                                        std::move(bound_constant));
+		expressions->push_back(std::move(filter_expr));
+		unique_values.insert(value);
 	}
-	if (children[0]->filter_type != TableFilterType::OPTIONAL_FILTER) {
+}
+
+void ExtractInFilterValues(ConjunctionAndFilter &filter, vector<Value> &values) {
+	if (filter.child_filters.empty()) {
 		return;
 	}
 
-	auto &optional_filter = children[0]->Cast<OptionalFilter>();
-	auto &child = optional_filter.child_filter;
-	if (child->filter_type != TableFilterType::IN_FILTER) {
+	// Extract the IN filter and the comparisons.
+	optional_ptr<InFilter> in_filter;
+	vector<reference<ConstantFilter>> comparisons;
+	for (idx_t i = 0; i < filter.child_filters.size(); i++) {
+		if (filter.child_filters[i]->filter_type == TableFilterType::CONSTANT_COMPARISON) {
```

Comment:
This seems pretty heavily tailored towards the filters that happen to be generated by the hash join.

* There's no reason to treat `OPTIONAL(IN_FILTER)` differently from `OPTIONAL(ANY_OTHER_FILTER)` - we should handle `OPTIONAL` filters generically
* We should handle `IN_FILTER` that does not occur in an `OPTIONAL` as well
* We should correctly handle multiple `IN_FILTERS` (either by aborting if we encounter multiple or intersecting them)